core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java

```diff
@@ -429,6 +562,68 @@ public <T extends RESTResponse> T handleRequest(
     return null;
   }
 
+  /**
+   * Do all the planning upfront but batch the file scan tasks across plan tasks. Plan Tasks have a
+   * key like <plan ID - table UUID - plan task sequence> The current implementation simply uses
+   * plan tasks as a pagination mechanism to control response sizes.
+   *
+   * @param tableScan
+   * @param planId
+   */
+  private void planFilesFor(TableScan tableScan, String planId) {
+    Iterable<List<FileScanTask>> taskGroupings =
+        Iterables.partition(
+            tableScan.planFiles(), planningBehavior.numberFileScanTasksPerPlanTask());
+    int planTaskSequence = 0;
+    String prevPlanTask = null;
+    for (List<FileScanTask> taskGrouping : taskGroupings) {
+      String planTaskKey =
+          String.format("%s-%s-%s", planId, tableScan.table().uuid(), planTaskSequence++);
+      planTaskToFileScanTasks.put(planTaskKey, taskGrouping);
+      if (prevPlanTask != null) {
+        planTaskToNext.put(prevPlanTask, planTaskKey);
+      }
+
+      prevPlanTask = planTaskKey;
+    }
+  }
+
+  private void asyncPlanFiles(TableScan scan, String asyncPlanId) {
+    asyncPlanningPool.submit(
+        () -> {
+          planFilesFor(scan, asyncPlanId);
+        });
+  }
+
+  // The initial set of file scan tasks is going to have a sentinel plan task which ends in
+  // 0. Directly return this set of file scan tasks as the initial set, along with
+  // any next plan task if applicable
+  private Pair<List<FileScanTask>, String> initialScanTasksForPlan(String planId) {
+    Set<Map.Entry<String, List<FileScanTask>>> initialPlanTaskAndFileScanTasks =
+        planTaskToFileScanTasks.entrySet().stream()
+            .filter(
+                planTask -> planTask.getKey().contains(planId) && planTask.getKey().endsWith("0"))
```

Comment:
should we split at the last `-` and see if its just 0, mostly coming from case where we have like `10` tasks